assembly/lab4ex1changliu.s

```diff
@@ -0,0 +1,308 @@
+    sw $s1, 8($sp)
+    sw $s2, 12($sp)
+    sw $s3, 16($sp)
+    sw $s4, 20($sp)
+
+    li $s0, 4
+    move $s2, $a0
+
+    print_matrix_loop:
+        la $a0, prompt_row
+        jal _print_str
+
+        # print row index
+        li $s1, 5
+        sub $s1, $s1, $s0
+        move $a0, $s1
+        jal _print_int
+
+        la $a0, prompt_row2
+        jal _print_str
+
+        # calculate array offset
+        addiu $t0, $s1, -1 # row 1 => offset 0
+        sll $t0, $t0, 4    # 4 words = 16 bytes = 2^4
+        addu $s3, $s2, $t0
+
+        # print first integer
+        lw $a0, 0($s3)
+        jal _print_int
+
+        # print rest with spaces
+        li $s4, 3
+        print_matrix_inner_loop:
+            la $a0, space
+            jal _print_str
+
+            addiu $s3, $s3, 4
+            lw $a0, 0($s3)
+            jal _print_int
+
+            addiu $s4, $s4, -1
+            bne $s4, $zero, print_matrix_inner_loop
+
+        # print newline
+        la $a0, newline
+        jal _print_str
+
+        addiu $s0, $s0, -1
+        bne $s0, $zero, print_matrix_loop
+
+    lw $s4, 20($sp)
+    lw $s3, 16($sp)
+    lw $s2, 12($sp)
+    lw $s1, 8($sp)
+    lw $s0, 4($sp)
+    lw $ra, 0($sp)
+    addiu $sp, $sp, 24
+
+    jr $ra
+
+# parameters:
+# $a0: (int *) pointer to result matrix
+# $a1: (int *) pointer to first matrix
+# $a2: (int *) pointer to second matrix
+multiply_matrix:
+    addiu $sp, $sp, -12
+    sw $s0, 0($sp)
+    sw $s1, 4($sp)
+    sw $s2, 8($sp)
+
+    # pseudo-code
+    # for (int i = 0; i < 4; ++i) {
+    #   for (int j = 0; j < 4; ++j) {
+    #     for (int k = 0; k < 4; ++k) {
+    #       $a0[i][j] += $a1[i][k] * $a2[k][j];
+    #     }
+    #   }
+    # }
+    move $s0, $a0
+    move $s1, $a1
+    move $s2, $a2
+
+    li $t3, 4
+
+    multiply_matrix_loop1:
+        addiu $t3, $t3, -1
+        li $t4, 4
+
+        multiply_matrix_loop2:
+            addiu $t4, $t4, -1
+            li $t5, 4
+
+            multiply_matrix_loop3:
+                addiu $t5, $t5, -1
+
+                # one row: 4 * 4 = 16 bytes
+                # one element: 4 bytes
+
+                # load
+                sll $t6, $t3, 4
+                sll $t7, $t5, 2
+                addu $t1, $t6, $t7
+                addu $t1, $t1, $s1
+                lw $t1, 0($t1)
+
+                sll $t6, $t5, 4
+                sll $t7, $t4, 2
+                addu $t2, $t6, $t7
+                addu $t2, $t2, $s2
+                lw $t2, 0($t2)
+
+                # multiply
+                mul $t1, $t1, $t2
+
+                # add in place
+                sll $t6, $t3, 4
+                sll $t7, $t4, 2
+                addu $t0, $t6, $t7
+                addu $t0, $t0, $s0
+                lw $t2, 0($t0)
+
+                addu $t2, $t2, $t1
+                sw $t2, 0($t0)
+
+                bne $t5, $zero, multiply_matrix_loop3
+
+            bne $t4, $zero, multiply_matrix_loop2
+
+        bne $t3, $zero, multiply_matrix_loop1
+
```

Comment:
I'm not sure whether this part would be ok or not. I didn't change the code on the book (actually I didn't find it). Instead I rewrote it myself.